### DISCUSSION TITLE
Generating a new access token server side

### DISCUSSION THREAD

**Discussion by joshsutheran-habitude** [supabase/supabase#18694] [Questions] [answered]
Hi all,

Due to the nature of our application, we are required to run authenticated API actions server side. These actions may occur outside of the standard hour duration that the JWT is valid for. 

My understanding of this was that I would be able to use the refresh token to handle this situation, though after a lot of time trying, cannot seem to find a solution.

Specifically, I want to be able to generate a new access token server side, using a refresh token that has been generated client side when the user authenticated.

Thanks,

Josh

**Comment by GaryAustin1** [comment]
You can use refreshSession with an refresh token to get a new session.  But then the client can no longer use that token and would need to be updated with the new token and jwt.

**Reply by GaryAustin1** [reply]
The other thing you can so server side is just mint a jwt with the Supabase secret and fill in the needed claims including the UUID in the sub claim.  That jwt expire can be set far out or minted each call.

**Reply by GaryAustin1** [reply]
Better not to generate new answers when responding to a suggestion.

When you createClient on a server you have to set persistSession:false or several operations will null the session out because there is no local storage.

**Reply by joshsutheran-habitude** [reply]
Apologies for the answer. Thanks for the above. If I understand correctly, the following should generate a new, valid access token when called server side (assuming the refresh token is valid)

```
const accessToken = sessionObj.access_token;
const refresh = sessionObj.refresh_token;
const options = {
  persistSession: false,
};

const supabase = createClient(supabaseUrl, anonKey, options);

const { data, error } = await supabase.auth.refreshSession({ refresh });
console.log(data, error);
```

The above does not work though - I am returned the following for data and error:
{ user: null, session: null } AuthSessionMissingError: Auth session missing!

**Reply by GaryAustin1** [reply]
Your persistSession is not being done correctly.
![image](https://github.com/supabase/supabase/assets/54564956/503fa0f4-2a7c-4d70-8ced-5c64a974df6e)

Also log out your refresh_token to make sure it looks right.

**Reply by joshsutheran-habitude** [reply]
I still have the same response using this code:

```
const accessToken = sessionObj.access_token;
const refresh = sessionObj.refresh_token;
const options = {
  auth: {
    autoRefreshToken: false,
    persistSession: false,
  },
};

console.log(refresh);

const supabase = createClient(supabaseUrl, anonKey, options);

const { data, error } = await supabase.auth.refreshSession({ refresh });
console.log(data, error);
```

Refresh token is correct. 

Response:
{ user: null, session: null } AuthSessionMissingError: Auth session missing!

**Reply by GaryAustin1** [reply]
Got nothing else for you.

**Reply by joshsutheran-habitude** [reply]
No worries - thanks for your help Gary. Have submitted a ticket so will see what comes of it!

**Reply by GaryAustin1** [reply, marked as answer]
You know you are sending in {refresh:value}  You need to send in {refresh_token:value}.  It is needs to look like a session object just missing the access_token.

**Reply by joshsutheran-habitude** [reply]
Hi Gary,

Thanks again for your advice - I can confirm that this was indeed the issue. I would point out that the docs don't suggest that's the case? I'm relatively new to the software world and in particular Supabase, but the following: 

```
const { data, error } = await supabase.auth.refreshSession({ refresh_token })
const { session, user } = data
```

does not suggest to me that refresh_token is a key, with the refresh token expected as a value. 

Thanks for your help! 👍

**Reply by GaryAustin1** [reply]
You can make doc improvement suggestions to them in Supabase/Supabase issues
